tox.ini

```diff
@@ -23,12 +23,19 @@ commands =
     twine check dist/*
 
 [testenv:release]
-basepython = python3
+skip_install = True
 deps =
-    wheel
+    pep517>=0.5
+    twine>=1.13
+    path.py
+passenv =
+    TWINE_PASSWORD
+setenv =
+    TWINE_USERNAME = {env:TWINE_USERNAME:__token__}
 commands =
-    python setup.py -q bdist_wheel sdist
-    twine upload --skip-existing dist/*
+    python -c "import path; path.Path('dist').rmtree_p()"
+    python -m pep517.build .
```

Comment:
It seems like now we'll have the tox environment install phase and the `pep517` install phase for the build requirements. This seems suboptimal. Why is packaging getting worse in Python?

Reply:
You have the tox install phase to handle the orchestration (incl. release) and the `pep517` phase (actually two, one for sdist, one for wheel) for the isolated builds. It's actually an improvement because now the environments are isolated, so dependencies in the orchestration don't affect the builds (and vice versa), and concerns are separated.

It was convenient when all of the tooling (prerequisites, dependencies, build, upload) was available in one tool (setuptools), but that monolithic design was complicated and to manage that complication, the tooling is being broken out into different tools (like pep517 and twine), but of course that means that something needs to come in to handle the orchestration that was formerly implied by having a single tool.